File: Provenance/Controller/iCade/PViCade8BitdoController.m
//
//  PViCade8BitdoController.m
//  Provenance
//
//  Created by Josejulio Martínez on 10/07/15.
//  Copyright (c) 2015 Josejulio Martínez. All rights reserved.
//

#import "PViCade8BitdoController.h"
#import "PViCadeGamepad.h"
#import "PViCadeGamepadButtonInput.h"
#import "PViCadeGamepadDirectionPad.h"

@implementation PViCade8BitdoController

-(instancetype) init {
    if (self = [super init]) {
        __unsafe_unretained PViCadeController* weakSelf = self;
        self.reader.buttonDown = ^(iCadeState button) {
            switch (button) {
                case iCadeButtonA:
                    [[weakSelf.iCadeGamepad buttonX] buttonPressed];
                    break;
                case iCadeButtonB:
                    [[weakSelf.iCadeGamepad buttonA] buttonPressed];
                    break;
                case iCadeButtonC:
                    [[weakSelf.iCadeGamepad buttonB] buttonPressed];
                    break;
                case iCadeButtonD:
                    [[weakSelf.iCadeGamepad buttonY] buttonPressed];
                    break;
                case iCadeButtonE:
                    [[weakSelf.iCadeGamepad rightShoulder] buttonPressed];
                    break;
                case iCadeButtonF:
                    [[weakSelf.iCadeGamepad leftShoulder] buttonPressed];
                    break;
                case iCadeButtonG:
                    [[weakSelf.iCadeGamepad leftTrigger] buttonPressed];
                    break;
                case iCadeButtonH:
                    [[weakSelf.iCadeGamepad rightTrigger] buttonPressed];
                    break;
                case iCadeJoystickDown:
                case iCadeJoystickLeft:
                case iCadeJoystickRight:
                case iCadeJoystickUp:
                    [[weakSelf.iCadeGamepad dpad] padChanged];
                    break;
                default:
                    break;
            }
            if (weakSelf.controllerPressedAnyKey) {
                weakSelf.controllerPressedAnyKey(weakSelf);
            }
        };
        
        self.reader.buttonUp = ^(iCadeState button) {
            switch (button) {
                case iCadeButtonA:
                    [[weakSelf.iCadeGamepad buttonX] buttonReleased];
                    break;
                case iCadeButtonB:
                    [[weakSelf.iCadeGamepad buttonA] buttonReleased];
                    break;
                case iCadeButtonC:
                    [[weakSelf.iCadeGamepad buttonB] buttonReleased];
                    break;
                case iCadeButtonD:
                    [[weakSelf.iCadeGamepad buttonY] buttonReleased];
                    break;
                case iCadeButtonE:
                    [[weakSelf.iCadeGamepad rightShoulder] buttonReleased];
                    break;
                case iCadeButtonF:
                    [[weakSelf.iCadeGamepad leftShoulder] buttonReleased];
                    break;
                case iCadeButtonG:
                    [[weakSelf.iCadeGamepad leftTrigger] buttonReleased];
                    break;
                case iCadeButtonH:
                    [[weakSelf.iCadeGamepad rightTrigger] buttonReleased];
                    break;
                case iCadeJoystickDown:
                case iCadeJoystickLeft:
                case iCadeJoystickRight:
                case iCadeJoystickUp:
                    [[weakSelf.iCadeGamepad dpad] padChanged];
                    break;
                default:
                    break;
            }
        };
    }
    return self;
}

- (NSString *) vendorName {
    return @"8Bitdo";
}

@end

File: PVSupport/Controller/iCade/Controllers/PViCade8BitdoController.swift
//  Converted to Swift 4 by Swiftify v4.2.29618 - https://objectivec2swift.com/
//
//  PViCade8BitdoController.swift
//  Provenance
//
//  Created by Josejulio Martínez on 10/07/15.
//  Copyright (c) 2015 Josejulio Martínez. All rights reserved.
//

import Foundation

public final class PViCade8BitdoController: PViCadeController {

    override func button(forState button: iCadeControllerState) -> PViCadeGamepadButtonInput? {
        switch button {
        case iCadeControllerState.buttonA:
            return self.iCadeGamepad.buttonX
        case iCadeControllerState.buttonB:
            return self.iCadeGamepad.buttonA
        case iCadeControllerState.buttonC:
            return self.iCadeGamepad.buttonB
        case iCadeControllerState.buttonD:
            return self.iCadeGamepad.buttonY
        case iCadeControllerState.buttonE:
            return self.iCadeGamepad.rightShoulder
        case iCadeControllerState.buttonF:
            return self.iCadeGamepad.leftShoulder
        case iCadeControllerState.buttonG:
            return self.iCadeGamepad.rightTrigger
        case iCadeControllerState.buttonH:
            return self.iCadeGamepad.leftTrigger
        default:
            return nil
        }
    }

    public override var vendorName: String? {
        return "8Bitdo"
    }
}

public final class PViCade8BitdoSNES30Controller: PViCadeController {

    override func button(forState button: iCadeControllerState) -> PViCadeGamepadButtonInput? {
        switch button {
        case iCadeControllerState.buttonA:
            return self.iCadeGamepad.buttonX
        case iCadeControllerState.buttonB:
            return self.iCadeGamepad.buttonA
        case iCadeControllerState.buttonC:
            return self.iCadeGamepad.buttonB
        case iCadeControllerState.buttonD:
            return self.iCadeGamepad.buttonY
        case iCadeControllerState.buttonE:
            return self.iCadeGamepad.rightShoulder
        case iCadeControllerState.buttonF:
            return self.iCadeGamepad.leftShoulder
        case iCadeControllerState.buttonG:
            return self.iCadeGamepad.rightTrigger
        case iCadeControllerState.buttonH:
            return self.iCadeGamepad.leftTrigger
        default:
            return nil
        }
    }

    public override var vendorName: String? {
        return "8Bitdo SNES30"
    }
}

public final class PViCade8BitdoZeroController: PViCadeController {

    override func button(forState button: iCadeControllerState) -> PViCadeGamepadButtonInput? {
        switch button {
        case iCadeControllerState.buttonA:
            return self.iCadeGamepad.buttonY
        case iCadeControllerState.buttonB:
            return self.iCadeGamepad.buttonB
        case iCadeControllerState.buttonC:
            return self.iCadeGamepad.buttonA
        case iCadeControllerState.buttonD:
            return self.iCadeGamepad.buttonX
        case iCadeControllerState.buttonE:
            return self.iCadeGamepad.rightShoulder
        case iCadeControllerState.buttonF:
            return self.iCadeGamepad.leftShoulder
        case iCadeControllerState.buttonG:
            return self.iCadeGamepad.rightTrigger
        case iCadeControllerState.buttonH:
            return self.iCadeGamepad.leftTrigger
        default:
            return nil
        }
    }

	override public var vendorName: String? {
        return "8Bitdo Zero"
    }
}
